crates/ruff/src/jupyter/index.rs

```diff
@@ -0,0 +1,96 @@
+use std::iter;
+use std::sync::Arc;
+
+use ruff_newlines::NewlineWithTrailingNewline;
+
+use crate::jupyter::{Cell, SourceValue};
+
+/// Jupyter Notebook indexing table
+///
+/// When we lint a jupyter notebook, we have to translate the row/column based on
+/// [`ruff_text_size::TextSize`] to jupyter notebook cell/row/column.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct JupyterIndex {
+    pub(super) inner: Arc<JupyterIndexInner>,
```

Comment:
So, `Diagnostics` takes ownership of `JupyterIndex` once the linting is completed to emit the correct information for each diagnostic w.r.t a Jupyter notebook. Now, that the index belongs to `Notebook` struct and cloning would be expensive, `Arc` made sense to me. On the other hand, we could take a reference and use lifetimes on `Diagnostics`.

Reply:
We can avoid using `Arc` altogether by storing the `SourceKind` on `Diagnostics` instead of `JupyterIndex`. But, this would mean that `SourceKind` (thus, `Notebook` and all of the schema types) need to implement the `PartialEq` and `Eq` trait.